compatibility_server/pip_checker.py

```diff
@@ -197,20 +199,54 @@ def __init__(self,
         self._tmp_path = tmp_path
         self._output_directory = None
         self._clean = clean
-
-    @staticmethod
-    def _run_command(command, stdout_path, stderr_path, raise_on_failure=True):
-        with open(stdout_path, 'w') as stdout_file, open(stderr_path,
-                                                         'w') as stderr_file:
-            completed_command = subprocess.run(
-                command, stdout=stdout_file, stderr=stderr_file)
-
-        logging.debug('Running %s [returncode=%s]', command,
-                      completed_command.returncode)
-        if completed_command.returncode and raise_on_failure:
-            raise PipError(command, completed_command.returncode, stderr_path)
-
-        return completed_command.returncode
+        self._docker_client = docker.from_env()
+
+    def _get_base_image(self):
+        python_version = self._pip_command[0]
+
+        if python_version == 'python2':
+            base_image = "python:2.7"
+        elif python_version == 'python3':
+            base_image = "python:3.6"
+
+        return base_image
+
+    def _build_container(self,
+                         container_name,
+                         command,
+                         stderr=True,
+                         detach=True):
+        container = self._docker_client.containers.run(
+            container_name,
+            command=command,
+            stderr=stderr,
+            detach=detach)
+        result = container.wait()
+        returncode = result.get('StatusCode')
+        return container, returncode
+
+    def _run_command(self, command, stdout_path, stderr_path):
```

Comment:
Your approach of saving and restoring the docker container is too clever ;-)

I think that it would be simpler (and possibly much faster) to do something like (this is just conceptual):

```python

def run()
  self._container = self._docker_client.containers.run(
            container_name,
            command='sleep 300', # exit if not done in 5 minutes, that's all of our CPU that you can use for bitcoin ;-)
            detach=True)
  # More stuff

def _run_command(command, stdout_path, stderr_path, raise_on_failure=True):
  exit_code, _ = self._container.exec_run(command)
  ...
```

Reply:
Good suggestion! I replaced the container commit with exec_run and it's much faster!

Reply:
![screenshot from 2018-11-07 13-11-09](https://user-images.githubusercontent.com/12888824/48161108-a0742600-e28e-11e8-9ead-51bbcb0b75df.png)